### PR TITLE
 WINNF_FT_S_FPR5  testcase

### PR DESCRIPTION
**Pull request by jcereced**
Create WINNF_FT_S_FPR_testcase.py

### CODE REVIEW

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        'fccId': device_1['fccId'],
+        'cbsdSerialNumber': device_1['cbsdSerialNumber'],
+        'airInterface': device_1['airInterface'],
+        'installationParam': device_1['installationParam'],
+        'measCapability': device_1['measCapability']
+    }
+    del device_1['cbsdCategory']
+    del device_1['airInterface']
+    del device_1['installationParam']
+
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: Also delete measCapability?

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    # Create the actual config.
+    conditionals = {
+        'registrationData': [conditionalParameters]
+    }
+    config = {
+      'registrationRequests': [device_1, device_2, device_3],
+      'conditionalRegistrationData': conditionals,
+      'grantRecords': [grant_1, grant_2, grant_3],
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: To be consistent with other PRs: change 'grantRecords' to 'grantRequests'

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      'gwblRecords': [gwbl],
+      'fssRecords': [fss]
+    }
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: We only need one GWBL and one FSS, so no need to make these plural and arrays.

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    writeConfig(filename, config)
+
+  @configurable_testcase(generate_FPR_5_default_config)
+  def test_WINNF_FT_S_FPR_5(self, config_filename):
+    """[Configurable] Grant Requests from one or more CBSDs Inside an FSS-GWBL Exclusion Zone."""
+
+    config = loadConfig(config_filename)
+        
+    # Light checking of the config file
+    self.assertEqual( len(config['registrationRequests']), 3)
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: Why are we checking that its exactly 3? (It does not say that in the test spec)

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    self.assertEqual( len(config['fssRecords']), 1)
+    self.assertEqual( len(config['gwblRecords']), 1)
+
+    # Load the FSS
+    self._sas_admin.InjectFss({'record': config['fssRecords']})
+
+    # Load the GWBL
+    self._sas_admin.InjectWisp(config['gwblRecords'])
+
+    # Trigger CPAS activity if GWBL > 0
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: This doesn't match the test spec. We should always trigger.

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    # Register N > 0 CBSDs
+    # Check registration response
+    # The assertRegistered function does the Inject FCC ID and user ID for the registration requests
+    if ('conditionalRegistrationData' in config) and (config['conditionalRegistrationData']):
+      cbsd_ids = self.assertRegistered(config['registrationRequests'],config['conditionalRegistrationData'])
+    else:
+      cbsd_ids = self.assertRegistered(config['registrationRequests'])
+
+    # number of grants and devices should match
+    self.assertEqual(len(config['registrationRequests']), len(config['grantRecords']))
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: Lines 145, 146, can be grouped along with the initial checking of the config file.

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+    # Add cbsdIds to grants
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: Lines 149 to 155 can be replaced by 'addCbsdIdsToRequests' helper function:
grant_request = config['grantRequests']
addCbsdIdsToRequests(cbsd_ids, grant_request)

(addCbsdIdsToRequests to be imported in Line 20)

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        'airInterface': device_1['airInterface'],
+        'installationParam': device_1['installationParam'],
+        'measCapability': device_1['measCapability']
+    }
+    del device_1['cbsdCategory']
+    del device_1['airInterface']
+    del device_1['installationParam']
+
+    # Create the actual config.
+    conditionals = {
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: conditionalRegistrationData changes coming up in: PR #479, so here too, conditionals need to be just a list of registration requests. And we wont need the checks in Line 140.

**Comment by jcereced** [comment]
Thanks!  We'll get them fix and let you know when it's ready.

**Comment by jcereced** [comment]
Changes have been made.  Please re-review.  thanks!

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+  def tearDown(self):
+    pass
+    
+  def generate_FPR_5_default_config(self, filename):
+    """Generates the WinnForum configuration for FPR.5."""
+
+    # Load FSS station operating at range 3600 - 3700 MHz.
+    fss = json.load(
+        open(os.path.join('testcases', 'testdata', 'fss_record_0.json')))
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: Please refer the changes to fss_record_0.json in PR#486. 

So it needs to be:
fss_data = json.load(open(os.path.join('testcases', 'testdata', 'fss_record_0.json')))
fss =  fss_data['record']

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+  def test_WINNF_FT_S_FPR_5(self, config_filename):
+    """[Configurable] Grant Requests from one or more CBSDs Inside an FSS-GWBL Exclusion Zone."""
+
+    config = loadConfig(config_filename)
+        
+    # Light checking of the config file
+    self.assertEqual(len(config['registrationRequests']), len(config['grantRequests']))
+
+    # Load the FSS
+    self._sas_admin.InjectFss({'record': config['fssRecord']})
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: self._sas_admin.InjectFss(config['fssRecord'])

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    # Load the GWBL
+    self._sas_admin.InjectWisp(config['gwblRecord'])
+
+    # Trigger CPAS activity
+    self.TriggerDailyActivitiesImmediatelyAndWaitUntilComplete() 
+
+    # Register N > 0 CBSDs
+    # Check registration response
+    # The assertRegistered function does the Inject FCC ID and user ID for the registration requests
+    if ('conditionalRegistrationData' in config) and (config['conditionalRegistrationData']):
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: We can just have: 

if ('conditionalRegistrationData' in config):
  cbsd_ids = self.assertRegistered(config['registrationRequests'],config['conditionalRegistrationData'])

File: src/harness/testcases/WINNF_FT_S_FPR_testcase.py [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    del device_1['measCapability']
+
+    # Create the actual config.
+    conditionals = [conditionalParameters]
+    config = {
+      'registrationRequests': [device_1, device_2, device_3],
+      'conditionalRegistrationData': conditionals,
+      'grantRequests': [grant_1, grant_2, grant_3],
+      'gwblRecord': gwbl,
+      'fssRecord': fss
```
Comment (harinineelu) [inline src/harness/testcases/WINNF_FT_S_FPR_testcase.py]: Here, it needs to be: 'fssRecord': 'fss_data'

**Comment by jcereced** [comment]
Made the requested changes.  please review.  Thanks!!

**Pull request merged**